inferno-ml-server-types/src/Inferno/ML/Server/Types.hs

```diff
@@ -0,0 +1,762 @@
+--
+-- Note that this is significantly more restrictive than Inferno's @Value@ type,
+-- which cannot have sensible @ToJSON@ and @FromJSON@ instances
+data IValue
+  = IText Text
+  | IDouble Double
+  | ITuple (IValue, IValue)
+  | ITime EpochTime
+  | IEmpty
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData)
+
+instance FromJSON IValue where
+  parseJSON = \case
+    String t -> pure $ IText t
+    Number n -> pure . IDouble $ toRealFloat n
+    -- It's easier to just mark the time explicitly in an object,
+    -- rather than try to deal with distinguishing times and doubles
+    Object o -> ITime <$> o .: "time"
+    Array a
+      | [x, y] <- Vector.toList a ->
+          fmap ITuple $
+            (,) <$> parseJSON x <*> parseJSON y
+      | Vector.null a -> pure IEmpty
+    _ -> fail "Expected one of: string, double, time, tuple, unit (empty array)"
+
+instance ToJSON IValue where
+  toJSON = \case
+    IDouble d -> toJSON d
+    IText t -> toJSON t
+    ITuple t -> toJSON t
+    -- See `FromJSON` instance above
+    ITime t -> object ["time" .= t]
+    IEmpty -> toJSON ()
+
+-- | Used to represent inputs to the script. 'Many' allows for an array input
+data SingleOrMany a
+  = Single a
+  | Many (Vector a)
+  deriving stock (Show, Eq, Generic, Functor)
+  deriving anyclass (NFData)
+
+instance FromJSON a => FromJSON (SingleOrMany a) where
+  parseJSON v =
+    asum
+      [ Single <$> parseJSON v,
+        Many <$> parseJSON v
+      ]
+
+instance ToJSON a => ToJSON (SingleOrMany a) where
+  toJSON = \case
+    Single a -> toJSON a
+    Many as -> toJSON as
+
+instance Ord a => Ord (SingleOrMany a) where
+  compare a =
+    (a,) >>> \case
+      (Single x, Single y) -> compare x y
+      (Many xs, Many ys) -> compare xs ys
+      (Single _, Many _) -> LT
+      (Many _, Single _) -> GT
+
+tshow :: Show a => a -> Text
+tshow = Text.pack . show
+
+maybeConversion ::
+  Typeable b => (a -> Maybe b) -> Field -> Maybe a -> Conversion b
+maybeConversion f fld =
+  maybe (returnError UnexpectedNull fld mempty) $
+    maybe (returnError ConversionFailed fld mempty) pure
+      . f
+
+-- ISO63912 language tag for model card
+
+data ISO63912 = ISO63912
+  { code :: (Char, Char),
+    name :: Text
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (NFData)
+
+instance FromJSON ISO63912 where
+  parseJSON = withText "ISO63912" $ \case
+    t
+      | Just (f, r) <- Text.uncons t,
+        Just (s, l) <- Text.uncons r,
+        l == mempty ->
+          maybe (fail "Missing ISO6391 language") pure $ fromChars (f, s)
+      | otherwise -> fail "Invalid ISO63912 code"
+    where
+      fromChars :: (Char, Char) -> Maybe ISO63912
+      fromChars c = ISO63912 c <$> Map.lookup c languagesByCode
+
+instance ToJSON ISO63912 where
+  toJSON (ISO63912 (f, s) _) = String . flip Text.snoc s $ Text.singleton f
+
+{- ORMOLU_DISABLE -}
+languagesByCode :: Map (Char, Char) Text
+languagesByCode =
```

Comment:
I don't know if we actually need these, but it might be nice for an eventual frontend to only display valid ISO 6391-2 codes. (There's a package for this on hackage, but it's ancient and uses a type with 100+ constructors to represent the code)